RFC: RBD+CephFS fscrypt support #1563

irq0 opened the pull request:
Add fscrypt integration with the Ceph CSI KMS. Supports Ceph FS and ext4 on RBD storage classes.
Implementation is based on the design doc https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/cephfs-fscrypt.md.

Testing this is currently non trivial. ext4 requires kernel support (see https://github.com/google/fscrypt#runtime-dependencies; Ceph FS requires a custom built kernel and a Ceph MDS from main. Unfortunately minikube currently doesn't have neither.

k8s config and tools to run this with a vstart cluster are in https://github.com/irq0/dev-ceph-csi-fscrypt-config

irq0 committed:
util: Make encryption passphrase size a parameter

The CephFS fscrypt support requires keys longer than 20 bytes. As a
preparation, make the new passphrase length configurable, but default
to 20 bytes

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
kms: Add GetSecret() to metadata KMS

Add GetSecret() to allow the fscrypt integration to access passphrases
directly without KDF and wrapping by a DEKStore.

This is intended to allow users to take a k8s secret, for example, and
use that directly as a password in fscrypt.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
kms: Add KMS test dummies

Let KMS implementations provide a CreateTestDummy() function returning
an EncryptionKMS just like Initializer, but without requiring
additional configuration and meant only for unit tests.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
cephfs: Add rudimentary fscrypt passphrase from KMS test

Add rudimentary test to ensure that we can get a valid passphrase from
the GetSecret feature

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: Refactor (block) encryption support in prep for file encryption

In preparation of fscrypt support for RBD filesystems, rename block
encryption related function to include the word 'block'. Add struct
fields and IsFileEncrypted in preparation for the fscrypt integration.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: Move fscrypt code from cephfs to util

Move fscrypt init, unlock and utility code form the cephfs package to
the common util package in preparation of RBD support

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: basic fscrypt file encryption support

Integrate basic fscrypt functionality into RBD initialization. To
activate file encryption instead of block introduce the new
'encryptionType' storage class key.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: fscrypt: FS specific checks, check encrypted file attrib

Add filesystem specific checking (cephfs exports an extra xattr) and
check for the encrypted inode attribute using ioctl_iflags(2).

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

humblec commented:
Thanks @irq0 !

nixpanic commented on go.mod:
```diff
@@ -42,6 +43,11 @@ require (
 	sigs.k8s.io/controller-runtime v0.11.0-beta.0.0.20211208212546-f236f0345ad2
 )
 
+require (
```
I do not understand why this is placed in a new `require` block. If it is done by tooling :shrug: , otherwise it should be fine to just have it combined in other blocks.

nixpanic commented on vendor/modules.txt:
```diff
@@ -212,6 +212,15 @@ github.com/golang/protobuf/ptypes/wrappers
 # github.com/golang/snappy v0.0.4
 ## explicit
 github.com/golang/snappy
+# github.com/google/fscrypt v0.3.3
```
you will also need to `go mod vendor` the updates and additions. Ideally in a separate commit, as that makes reviewing a little easier.

nixpanic commented on internal/kms/kms.go:
```diff
@@ -93,6 +93,15 @@ func GetKMS(tenant, kmsID string, secrets map[string]string) (EncryptionKMS, err
 	return kmsManager.buildKMS(tenant, kmsConfig, secrets)
 }
 
+func GetKMSTestDummy(kmsID string) EncryptionKMS {
```
if this is only used for unit testing, can it be placed in `kms_test.go` instead? Maybe add a
```go
type ProviderTest struct {
	Provider
	CreateTestDummy TestDummyFunc
}
```
in the `kms_test.go` for that as well? Not sure if that is sufficient though...

irq0 replied:
Good idea. I'll look into that.

irq0 replied:
Turned out a little bit different, but I managed to move the test dummy stuff out of the Provider struct

nixpanic commented on internal/cephfs/fscrypt_test.go:
```diff
@@ -0,0 +1,37 @@
+package cephfs
```
please include the licence/copyright header

nixpanic commented on internal/cephfs/fscrypt_test.go:
```diff
@@ -0,0 +1,37 @@
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/net/context"
```
use `"context"` instead?

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -105,10 +100,21 @@ func (ns *NodeServer) getVolumeOptions(
 	return volOptions, nil
 }
 
+func MaybeFscryptUnlock(ctx context.Context, volOptions *store.VolumeOptions,
```
New (exported) functions ideally have comments.

Does it make sense to have this as a method of `store.VolumeOptions`?

nixpanic commented on internal/util/fscrypt.go:
```diff
@@ -1,4 +1,4 @@
-package cephfs
+package util
```
The `util` package is really large already. If you can move this into `util/encryption` or something similar, please do!

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1469,7 +1469,7 @@ func (cs *ControllerServer) ControllerExpandVolume(
 	// Hence set nodeExpansion flag based on VolumeMode and Encryption status
 	nodeExpansion := true
-	if req.GetVolumeCapability().GetBlock() != nil && !rbdVol.isEncrypted() {
+	if req.GetVolumeCapability().GetBlock() != nil && !rbdVol.isBlockEncrypted() {
```
the commit message contains

> include the world 'block'

You probably mean s/world/word/

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -685,6 +699,17 @@ func (ns *NodeServer) NodePublishVolume(
 
+	fileEncrypted, err := util.IsEncrypted(req.GetVolumeContext())
+	if err != nil {
+		return nil, err
```
this function needs to return a gRPC error. Functions from the `util` package should only return 'plain' errors.

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -700,7 +725,8 @@ func (ns *NodeServer) mountVolumeToStagePath(
 	req *csi.NodeStageVolumeRequest,
 	staticVol bool,
-	stagingPath, devicePath string) error {
+	stagingPath, devicePath string, fileEncryption bool,
```
coding guidelines suggest to place new parameters of different types on the next line

nixpanic commented on internal/util/fscrypt.go:
```diff
@@ -210,15 +217,49 @@ func initializeAndUnlock(ctx context.Context, fscryptContext *fscryptactions.Con
+		if err != nil {
+			return fmt.Errorf("error reading ceph.fscrypt.auth xattr on %q: %w", directoryPath, err)
+		}
+	case "ext4":
```
maybe this should be the default case?

irq0 replied:
Makes sense, but let me double check if other filesystems with fscrypt support set the inode encryption flag.

irq0 replied:
Checked: ext4 and f2fs set the inode encrypted flag. Ceph FS doesn't support inode flags. Changing the ext4 to the default case.

nixpanic commented on internal/util/fscrypt.go:
```diff
@@ -210,15 +217,49 @@ func initializeAndUnlock(ctx context.Context, fscryptContext *fscryptactions.Con
-	if err != nil {
-		return fmt.Errorf("error reading ceph.fscrypt.auth xattr on %q: %w", directoryPath, err)
+	switch filesystem {
+	case "cephfs":
```
above you use `"ceph"`, here you check `"cephfs"`... Which is the right name? Might be better to make it a constant :wink:

irq0 replied:
Good catch! It should be "ceph", as that is what the kernel calls cephfs.

nixpanic commented on internal/util/fscrypt.go:
```diff
@@ -210,15 +217,49 @@ func initializeAndUnlock(ctx context.Context, fscryptContext *fscryptactions.Con
 	return nil
 }
 
+func getInodeEncryptedAttribute(p string) (bool, error) {
```
a little comment on what this function checks would be nice

nixpanic reviewed (commented):
Thanks for the PR! Some initial comments for now, I'll try to keep following the progress you're making.

Commit:
vendor: vendor fscrypt integration dependencies

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
cephfs: Refactor file encryption integration

Reduce cyclomatic complexity of NodeStageVolume by moving all related code
into maybeUnlockFileEncryption and maybeInitializeFileEncryption.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
cleanup: Fix nlreturn issues in kms/kms, rbd/nodeserver

Fix nlreturn linter issues raised by previous changes to the respective files.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: Determine best supported fscrypt policy on node init

Currently fscrypt supports policies version 1 and 2. 2 is the best
choice and was the only choice prior to this commit. Since we need to
support kernels < 5.4 (e.g minikube), select version 1 there.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
examples: Document new encryptionType storage class setting

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add fscrypt on rbd helper

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: Add util to fetch encryption type from vol options

Fetch encryption type from vol options. Make fallback type
configurable to support RBD (default block) and Ceph FS (default file)

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
cephfs: Parse encryption options similar to RBD

Use common utility code to parse encryption information from volume
options.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add helper to run encryption tests on block and file

Add a `By` wrapper to parameterize encryption related test functions
and run them on both block and file encryption

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Unlock: Fetch keys early

Fetch keys from KMS before doing anything else. This will catch KMS
errors before setting up any fscrypt metadata.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Fetch passphrase when keyFn is invoked not created

Fetch password when keyFn is invoked, not when it is created. This
allows creation of the keyFn before actually creating the passphrase.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Don't base64 decode passphrases from KMS

Commit:
fscrypt: Revert protector on error

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
journal: Store encryptionType next to KMS

Add encryptionType next to kmsID to support both block and file
encryption.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: Add Parse, ->String for EncryptionType

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
I renamed the branch. Apparently this closes attached PRs :). New PR is https://github.com/ceph/ceph-csi/pull/3310
